packages/flutter/lib/src/widgets/sliver.dart

```diff
@@ -966,35 +1010,78 @@ class SliverMultiBoxAdaptorElement extends RenderObjectElement implements Render
 
   final Map<int, Widget> _childWidgets = HashMap<int, Widget>();
   final SplayTreeMap<int, Element> _childElements = SplayTreeMap<int, Element>();
-  RenderBox _currentBeforeChild;
 
   @override
   void performRebuild() {
     _childWidgets.clear(); // Reset the cache, as described above.
     super.performRebuild();
-    _currentBeforeChild = null;
     assert(_currentlyUpdatingChildIndex == null);
     try {
+      final Set<int> affectedIndexes = Set<int>.from(_childElements.keys);
+      final Map<int, Element> newKeepAliveChildren = <int, Element>{};
+
+      int getNewIndex(Widget w){
+        if (w.key != null){
+          return widget.delegate.findIndexByKey(w.key);
+        } else {
+          return -1;
+        }
+      }
+
+      void removeChildIfNotRemapped(int index){
+        final Element child = _childElements[index];
+        final SliverMultiBoxAdaptorParentData originalParentData = child?.renderObject?.parentData;
+        if (originalParentData != null){
+          if (getNewIndex(child.widget) == -1) {
+            _childElements[index] = updateChild(_childElements[index], null, index);
+          }
+        }
+      }
+
       void processElement(int index) {
         _currentlyUpdatingChildIndex = index;
-        final Element newChild = updateChild(_childElements[index], _build(index), index);
+        Element newChild;
+        if (newKeepAliveChildren.containsKey(index)) {
+          if (_childElements[index] != null) {
+            removeChildIfNotRemapped(index);
+          }
+          newChild = updateChild(newKeepAliveChildren[index], _build(index), index);
+        } else {
+          newChild = updateChild(_childElements[index], _build(index), index);
+        }
         if (newChild != null) {
           _childElements[index] = newChild;
-          final SliverMultiBoxAdaptorParentData parentData = newChild.renderObject.parentData;
-          if (!parentData.keptAlive)
-            _currentBeforeChild = newChild.renderObject;
         } else {
           _childElements.remove(index);
         }
       }
-      // processElement may modify the Map - need to do a .toList() here.
-      _childElements.keys.toList().forEach(processElement);
+
+      // Remapping children if it is reordered.
+      for (int index in _childElements.keys.toList()) {
+        final SliverMultiBoxAdaptorParentData originalParentData = _childElements[index]?.renderObject?.parentData;
+        if (originalParentData != null) {
+          final int newIndex = getNewIndex(_childElements[index].widget);
+          if(newIndex!=-1 && newIndex != index){
+            newKeepAliveChildren[newIndex] = _childElements[index];
```

Comment:
newKeepAliveChildren -> newKeyedChildren? It's more than just the keep alive ones right?